### PR TITLE
feat(driver): support for init_module, finit_module syscalls

### PR DESCRIPTION
**Pull request by therealbobo**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

/area driver-kmod

/area driver-bpf

/area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

/area libscap

/area libpman

> /area libsinsp

/area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

/version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

This adds the support for init\_module, finit\_module syscalls.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/drivers/test_suites/syscall_enter_suite/finit_module_e.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	auto evt_test = get_syscall_event_test(__NR_finit_module, ENTER_EVENT);
+	evt_test->enable_capture();
+
+	/*=============================== TRIGGER SYSCALL  ===========================*/
+
+	int mock_fd = -1;
+	char mock_buf[8];
+	int mock_flags = 0;
+
+	assert_syscall_state(SYSCALL_SUCCESS, "finit_module", syscall(__NR_finit_module, mock_fd, (void *)(mock_buf), mock_flags));
```
Comment (Andreagit97) [inline test/drivers/test_suites/syscall_enter_suite/finit_module_e.cpp:14]:
```suggestion
	assert_syscall_state(SYSCALL_FAILURE, "finit_module", syscall(__NR_finit_module, mock_fd, (void *)(mock_buf), mock_flags));
```

File: test/drivers/test_suites/syscall_enter_suite/init_module_e.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	auto evt_test = get_syscall_event_test(__NR_init_module, ENTER_EVENT);
+	evt_test->enable_capture();
+
+	/*=============================== TRIGGER SYSCALL  ===========================*/
+
+	char mock_img[100];
+	unsigned long len = 100;
+	char mock_buf[8];
+
+	assert_syscall_state(SYSCALL_SUCCESS, "init_module", syscall(__NR_init_module, (void *)(mock_img), len, (void *)(mock_buf)));
```
Comment (Andreagit97) [inline test/drivers/test_suites/syscall_enter_suite/init_module_e.cpp:14]:
```suggestion
	assert_syscall_state(SYSCALL_FAILURE, "init_module", syscall(__NR_init_module, (void *)(mock_img), len, (void *)(mock_buf)));
```

File: test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	evt_test->assert_header();
+
+	/*=============================== ASSERT PARAMETERS  ===========================*/
+
+	/* Parameter 1: res (type: PT_ERRNO)*/
+	evt_test->assert_numeric_param(1, (uint64_t)errno_value);
+
+	/* Parameter 2: fd (type: PT_FD) */
+	evt_test->assert_numeric_param(2, (uint64_t)kmod_fd);
```
Comment (Andreagit97) [inline test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp:46]:
```suggestion
	evt_test->assert_numeric_param(2, (int64_t)kmod_fd);
```

File: test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+	/* Read data from /dev/urandom */
+	char mock_buf[] = "AAABAACAADAAEAAFAAGAAHAAIAAJAAKAALAAMAA\0";
+
+	/*
+	 * Call the `finit_module`
+	 */
+
+	int64_t kmod_fd = 99;
+	assert_syscall_state(SYSCALL_FAILURE, "finit_module", syscall(__NR_finit_module, kmod_fd, (void*)mock_buf, 1));
```
Comment (Andreagit97) [inline test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp:21]: Can we test all flags here? maybe using explicit `PPM_` notation

File: test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
+#include "../../event_class/event_class.h"
+
+#if defined(__NR_finit_module)
+
+#if defined(__NR_close) && defined(__NR_open) && defined(__NR_read)
```
Comment (Andreagit97) [inline test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp:5]: why do we need this `ifdef`?

File: test/drivers/test_suites/syscall_exit_suite/init_module_x.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	int fd = syscall(__NR_open, "/dev/urandom", O_RDONLY);
+	assert_syscall_state(SYSCALL_SUCCESS, "open", fd, NOT_EQUAL, -1);
+
+	/* Read data from /dev/urandom */
+	const unsigned data_len = DEFAULT_SNAPLEN / 2;
+	char mock_img[data_len];
+	ssize_t read_bytes = syscall(__NR_read, fd, (void *)mock_img, data_len);
+	assert_syscall_state(SYSCALL_SUCCESS, "read", read_bytes, NOT_EQUAL, 0);
+
+	char mock_buf[] = "AAABAACAADAAEAAFAAGAAHAAIAAJAAKAALAAMAA\0";
```
Comment (Andreagit97) [inline test/drivers/test_suites/syscall_exit_suite/init_module_x.cpp:23]: I would avoid `udev` here, i would use the same strategy used for `mock_buf` so just a `char` buffer + a random `len` in this way we can avoid extra syscalls, for example `open` is not supported on ARM64

**Review by Andreagit97 — COMMENTED** [review]
Amazing work!

**Comment by Andreagit97** [comment]
Please double-check these tests in sinsp:
```
[  FAILED  ] events_set.event_set_to_names_generic_events
[  FAILED  ] events_set.event_set_to_names_no_generic_events2
[  FAILED  ] interesting_syscalls.event_set_to_sc_set_generic_events
```
You will notice that they assert `INIT_MODULE` syscall as generic but from now on it won't be generic

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/finit_module.bpf.c [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	/* Parameter 2: fd (type: PT_FD) */
+	s32 fd = (s32)extract__syscall_argument(regs, 0);
+	auxmap__store_s64_param(auxmap, fd);
+
+	u16 snaplen = maps__get_snaplen();
+	apply_dynamic_snaplen(regs, &snaplen, false);
+
+	/* Parameter 3: uargs (type: PT_CHARBUF) */
+	unsigned long uargs_ptr = extract__syscall_argument(regs, 1);
+	auxmap__store_charbuf_param(auxmap, uargs_ptr, snaplen, USER);
```
[the remark this answers is not in the document]
Comment (therealbobo) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/finit_module.bpf.c:70]: Why should we use `MAX_PATH` here? `uargs` are the kernel module parameters as far as I know 🤔

File: test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp (hunk 5 above)
Comment (therealbobo) [reply test/drivers/test_suites/syscall_exit_suite/finit_module_x.cpp:5]: Ops, the test was using some random bytes from urandom, then I changed it.

File: driver/bpf/fillers.h [hunk 8]
```diff
@@ -7058,4 +7058,52 @@ FILLER(sys_pidfd_open_x, true)
...
+	CHECK_RES(res);
+
+	/* Parameter 3: uargs (type: PT_CHARBUF) */
+	long uargs = bpf_syscall_get_argument(data, 1);
+	res = bpf_val_to_ring(data, uargs);
+	CHECK_RES(res);
+
+	/* Parameter 4: flags (type: PT_INT32) */
+	s32 flags = bpf_syscall_get_argument(data, 2);
+	return bpf_push_s32_to_ring(data, finit_module_flags_to_scap(flags));
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply driver/bpf/fillers.h:7106]: Same in all other drivers

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**